rust/bin-shared/src/tun_device_manager/windows.rs

```diff
@@ -51,11 +71,79 @@ impl TunDeviceManager {
     }
 
     #[tracing::instrument(level = "trace", skip(self))]
-    pub async fn set_routes(&mut self, _: Vec<Ipv4Network>, _: Vec<Ipv6Network>) -> Result<()> {
-        // TODO: Windows still does route updates in `tun_windows.rs`. I can move it up
-        // here, but since the Client and Gateway don't know the index of the WinTun
-        // interface, I'd have to use the Windows API
-        // <https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/NetworkManagement/IpHelper/fn.GetAdaptersAddresses.html>
-        unimplemented!()
+    pub async fn set_routes(&mut self, v4: Vec<Ipv4Network>, v6: Vec<Ipv6Network>) -> Result<()> {
+        let iface_idx = self
+            .iface_idx
+            .context("Cannot set routes without having created TUN device")?;
+
+        let new_routes = HashSet::from_iter(
+            v4.into_iter()
+                .map(IpNetwork::from)
+                .chain(v6.into_iter().map(IpNetwork::from)),
+        );
+
+        if new_routes == self.routes {
+            return Ok(());
+        }
+
+        for new_route in new_routes.difference(&self.routes) {
+            add_route(*new_route, iface_idx)?;
+        }
+
+        for old_route in self.routes.difference(&new_routes) {
+            remove_route(*old_route, iface_idx)?;
+        }
+
+        self.routes = new_routes;
+
+        Ok(())
     }
 }
+
+// It's okay if this blocks until the route is added in the OS.
+fn add_route(route: IpNetwork, iface_idx: u32) -> Result<()> {
+    const DUPLICATE_ERR: u32 = 0x80071392;
+    let entry = forward_entry(route, iface_idx);
+
+    // SAFETY: Windows shouldn't store the reference anywhere, it's just a way to pass lots of arguments at once. And no other thread sees this variable.
+    match unsafe { CreateIpForwardEntry2(&entry) }.ok() {
+        Ok(()) => Ok(()),
+        Err(e) if e.code().0 as u32 == DUPLICATE_ERR => {
+            tracing::debug!(%route, "Failed to add duplicate route, ignoring");
+            Ok(())
+        }
+        Err(e) => Err(e.into()),
+    }
+}
+
+// It's okay if this blocks until the route is removed in the OS.
+fn remove_route(route: IpNetwork, iface_idx: u32) -> Result<()> {
+    let entry = forward_entry(route, iface_idx);
+
+    // SAFETY: Windows shouldn't store the reference anywhere, it's just a way to pass lots of arguments at once. And no other thread sees this variable.
+    unsafe { DeleteIpForwardEntry2(&entry) }.ok()?;
+    Ok(())
+}
+
+fn forward_entry(route: IpNetwork, iface_idx: u32) -> MIB_IPFORWARD_ROW2 {
+    let mut row = MIB_IPFORWARD_ROW2::default();
+    // SAFETY: Windows shouldn't store the reference anywhere, it's just setting defaults
+    unsafe { InitializeIpForwardEntry(&mut row) };
+
+    let prefix = &mut row.DestinationPrefix;
+    match route {
+        IpNetwork::V4(x) => {
+            prefix.PrefixLength = x.netmask();
```

Comment:
I know we're just moving this code from somewhere else, but I was scanning this and wanted to double-check -- this code is *not* simply assigning a netmask as a prefix length, correct?

Just making sure we don't need to do the [same conversion](https://github.com/firezone/firezone/blob/main/swift/apple/FirezoneNetworkExtension/NetworkSettings.swift#L65) we do in the Apple clients and this isn't a lurking bug.

Reply:
I honestly haven't reviewed this code but I'll take a look (and maybe write some tests for it).

Reply:
As far as I understand, this is correct. Here are the docs: https://learn.microsoft.com/en-us/windows/win32/api/netioapi/ns-netioapi-ip_address_prefix#members

It says the the `PrefixLength` is "The length, in bits, of the prefix or network part of the IP address.". That is equivalent to the netmask, i.e. If we have `192.168.1.0/24`, the netmask is `24`, indicating the first 24 bits of the address are the network part of the address.

Reply:
> Just making sure we don't need to do the [same conversion](https://github.com/firezone/firezone/blob/main/swift/apple/FirezoneNetworkExtension/NetworkSettings.swift?rgh-link-date=2024-07-14T22%3A13%3A42Z#L65) we do in the Apple clients and this isn't a lurking bug.

Btw, you'd have to special-case the `0` prefix but there shouldn't be a need for a lookup-table. I am not familiar with Swift's stdlib but in Rust, creating an IPv4 address from a u32 is a one-liner and calculating the correct subnet mask is a single bitshift operation:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=82998098bc845a5c7ff1ab3ab95534fe